### PR TITLE
add backward op for flatten and softmax; fix bug for softmax in dim 1 case

### PR DESCRIPTION
**Pull request by OscarXZQ**
加上了flatten和softmax的后向 (两者目前都没有动态属性)
修改了activation.py的transpose check，使softmax支持一维数组 （原本的softmax给入一维数组之后会报错）
加入了一个非特殊的test_softmax的样例

### CODE REVIEW

File: oneflow/python/test/modules/test_activation.py [hunk 1]
```diff
@@ -312,9 +312,10 @@ def _test_softmax_dim_3(test_case, device):
     test_case.assertTrue(np.allclose(y2.numpy(), output2, 1e-5, 1e-5))
 
 
-def _test_softmax_backward(test_case, device):
+def _test_softmax_backward_1(test_case, device):
```
Comment (BBuf) [inline oneflow/python/test/modules/test_activation.py:315]: 1改成nornal吧

File: oneflow/python/test/modules/test_activation.py [hunk 2]
```diff
@@ -329,6 +330,21 @@ def _test_softmax_backward(test_case, device):
     test_case.assertTrue(np.allclose(x.grad.numpy(), x_grad, 1e-5, 1e-5))
 
 
+def _test_softmax_backward_2(test_case, device):
```
Comment (BBuf) [inline oneflow/python/test/modules/test_activation.py:333]: 2改成1_dim

File: oneflow/python/nn/modules/activation.py [hunk 3]
```diff
@@ -24,7 +24,8 @@
 def _softmax_need_transpose(x, axis):
     assert type(axis) is int
     dim_num = len(x.shape)
-    assert dim_num >= 2
+    if dim_num <= 1:
```
Comment (BBuf) [inline oneflow/python/nn/modules/activation.py:27]: <去掉吧

File: oneflow/core/autograd/gradient_funcs/flatten.cpp [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+class Flatten : public OpExprGradFunction<FlattenInterpState> {
+ public:
+  Maybe<void> Init(const OpExpr& op) override;
+  Maybe<void> Capture(FlattenInterpState* ctx, const TensorTuple& inputs,
+                      const TensorTuple& outputs, const AttrMap& attrs) const override;
+  Maybe<void> Apply(const FlattenInterpState* ctx, const TensorTuple& out_grads,
+                    TensorTuple* in_grads) const override;
+
+ private:
+  AttrMap base_attrs_;
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/flatten.cpp:39]: 多余了

File: oneflow/core/autograd/gradient_funcs/flatten.cpp [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+  ctx->SaveTensorForBackward(inputs.at(0));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> Flatten::Apply(const FlattenInterpState* ctx, const TensorTuple& out_grads,
+                           TensorTuple* in_grads) const {
+  if (!ctx->requires_grad) { return Maybe<void>::Ok(); }
+  CHECK_EQ_OR_RETURN(out_grads.size(), 1);
+  const std::shared_ptr<oneflow::one::Tensor>& like = ctx->SavedTensors().at(0);
+  MutableAttrMap attrs;
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/flatten.cpp:66]: 删掉attrs

File: oneflow/core/autograd/gradient_funcs/flatten.cpp [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+  CHECK_NOTNULL_OR_RETURN(fw_op_expr);
+  base_attrs_ = MakeAttrMapFromUserOpConf(fw_op_expr->proto());
+  const std::string& op_name = fw_op_expr->op_name();
+  grad_op_ = JUST(op_expr_helper::ReshapeLikeOp(GradientOpName(op_name)));
+  return Maybe<void>::Ok();
+}
+
+Maybe<void> Flatten::Capture(FlattenInterpState* ctx, const TensorTuple& inputs,
+                             const TensorTuple& outputs, const AttrMap& attrs) const {
+  ComposedAttrMap composed_attrs(attrs, base_attrs_);
```
Comment (BBuf) [inline oneflow/core/autograd/gradient_funcs/flatten.cpp:54]: 也删掉

**Review by BBuf — APPROVED** [review]
